File: go/comparison/byline.go
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Go.

package comparison

import (
	"unicode"

	"github.com/byron1st/file-diff/go/myers"
	"github.com/byron1st/file-diff/go/util"
)

// Line wraps a line of text with its comparison policy, cached hash, and
// non-space character count.
type Line struct {
	Content       string
	Policy        ComparisonPolicy
	hash          int
	nonSpaceChars int
}

func NewLine(content string, policy ComparisonPolicy) *Line {
	return &Line{
		Content:       content,
		Policy:        policy,
		hash:          HashCode(content, policy),
		nonSpaceChars: countNonSpaceChars(content),
	}
}

func (l *Line) equals(other *Line) bool {
	if l == other {
		return true
	}
	if l.hash != other.hash {
		return false
	}
	return IsEqual(l.Content, other.Content, l.Policy)
}

// Equals implements lineEquatable for ExpandChangeBuilder.
func (l *Line) Equals(other lineEquatable) bool {
	if o, ok := other.(*Line); ok {
		return l.equals(o)
	}
	return false
}

func countNonSpaceChars(s string) int {
	n := 0
	for _, c := range s {
		if !unicode.IsSpace(c) {
			n++
		}
	}
	return n
}

// CompareLines is the main entry point for line-level comparison.
// It implements the ByLineRt.compare algorithm.
func CompareLines(lines1, lines2 []string, policy ComparisonPolicy) FairDiffIterable {
	l1 := toLines(lines1, policy)
	l2 := toLines(lines2, policy)
	return doCompareLines(l1, l2, policy)
}

func doCompareLines(lines1, lines2 []*Line, policy ComparisonPolicy) FairDiffIterable {
	if policy == PolicyIgnoreWhitespaces {
		changes := compareSmart(lines1, lines2)
		changes = optimizeLineChunks(lines1, lines2, changes)
		return expandRanges(lines1, lines2, changes)
	}

	iwLines1 := convertMode(lines1, PolicyIgnoreWhitespaces)
	iwLines2 := convertMode(lines2, PolicyIgnoreWhitespaces)

	iwChanges := compareSmart(iwLines1, iwLines2)
	iwChanges = optimizeLineChunks(lines1, lines2, iwChanges)
	return correctChangesSecondStep(lines1, lines2, iwChanges)
}

// compareSmart compares in two steps:
// 1. Compare ignoring "unimportant" (short) lines
// 2. Fill gaps using full comparison
func compareSmart(lines1, lines2 []*Line) FairDiffIterable {
	threshold := myers.UnimportantLineCharCount()
	if threshold == 0 {
		result, _ := diffLineSlice(lines1, lines2)
		return result
	}

	bigLines1, idx1 := getBigLines(lines1, threshold)
	bigLines2, idx2 := getBigLines(lines2, threshold)

	changes, err := diffLineSlice(bigLines1, bigLines2)
	if err != nil {
		// fallback to full diff
		result, _ := diffLineSlice(lines1, lines2)
		return result
	}

	return newSmartLineChangeCorrector(idx1, idx2, lines1, lines2, changes).build()
}

func getBigLines(lines []*Line, threshold int) ([]*Line, []int) {
	var big []*Line
	var indexes []int
	for i, l := range lines {
		if l.nonSpaceChars > threshold {
			big = append(big, l)
			indexes = append(indexes, i)
		}
	}
	return big, indexes
}

// correctChangesSecondStep adjusts IW-matched lines to prefer exact matches.
func correctChangesSecondStep(lines1, lines2 []*Line, changes FairDiffIterable) FairDiffIterable {
	eqLines1 := make([]lineEquatable, len(lines1))
	for i, l := range lines1 {
		eqLines1[i] = l
	}
	eqLines2 := make([]lineEquatable, len(lines2))
	for i, l := range lines2 {
		eqLines2[i] = l
	}
	builder := NewExpandChangeBuilder(eqLines1, eqLines2)

	var sample string
	hasSample := false
	last1, last2 := 0, 0

	for _, r := range changes.Unchanged() {
		count := r.End1 - r.Start1
		for i := range count {
			idx1 := r.Start1 + i
			idx2 := r.Start2 + i
			l1 := lines1[idx1]
			l2 := lines2[idx2]

			if !hasSample || !IsEqual(sample, l1.Content, PolicyIgnoreWhitespaces) {
				if l1.equals(l2) {
					flushSecondStep(builder, lines1, lines2, sample, hasSample, last1, last2, idx1, idx2)
					hasSample = false
					builder.MarkEqual(idx1, idx2)
				} else {
					flushSecondStep(builder, lines1, lines2, sample, hasSample, last1, last2, idx1, idx2)
					sample = l1.Content
					hasSample = true
				}
			}
			last1 = idx1 + 1
			last2 = idx2 + 1
		}
	}
	flushSecondStep(builder, lines1, lines2, sample, hasSample, last1, last2, changes.Length1(), changes.Length2())

	return Fair(builder.Finish())
}

func flushSecondStep(
	builder *ExpandChangeBuilder,
	lines1, lines2 []*Line,
	sample string, hasSample bool,
	last1, last2, line1, line2 int,
) {
	if !hasSample {
		return
	}

	start1 := max(last1, builder.Index1())
	start2 := max(last2, builder.Index2())

	var sub1, sub2 []int
	for i := start1; i < line1; i++ {
		if IsEqual(sample, lines1[i].Content, PolicyIgnoreWhitespaces) {
			sub1 = append(sub1, i)
		}
	}
	for i := start2; i < line2; i++ {
		if IsEqual(sample, lines2[i].Content, PolicyIgnoreWhitespaces) {
			sub2 = append(sub2, i)
		}
	}

	if len(sub1) == 0 || len(sub2) == 0 {
		return
	}

	alignExactMatching(builder, lines1, lines2, sub1, sub2)
}

func alignExactMatching(builder *ExpandChangeBuilder, lines1, lines2 []*Line, sub1, sub2 []int) {
	n := max(len(sub1), len(sub2))
	skipAligning := n > 10 || len(sub1) == len(sub2)

	if skipAligning {
		count := min(len(sub1), len(sub2))
		for i := range count {
			if lines1[sub1[i]].equals(lines2[sub2[i]]) {
				builder.MarkEqual(sub1[i], sub2[i])
			}
		}
		return
	}

	if len(sub1) < len(sub2) {
		matching := getBestMatchingAlignment(sub1, sub2, lines1, lines2)
		for i := range len(sub1) {
			if lines1[sub1[i]].equals(lines2[sub2[matching[i]]]) {
				builder.MarkEqual(sub1[i], sub2[matching[i]])
			}
		}
	} else {
		matching := getBestMatchingAlignment(sub2, sub1, lines2, lines1)
		for i := range len(sub2) {
			if lines1[sub1[matching[i]]].equals(lines2[sub2[i]]) {
				builder.MarkEqual(sub1[matching[i]], sub2[i])
			}
		}
	}
}

func getBestMatchingAlignment(shorter, longer []int, linesS, linesL []*Line) []int {
	size := len(shorter)
	best := make([]int, size)
	for i := range size {
		best[i] = i
	}

	comb := make([]int, size)
	bestWeight := 0

	var combinations func(start, n, k int)
	combinations = func(start, n, k int) {
		if k == size {
			weight := 0
			for i := range size {
				if linesS[shorter[i]].equals(linesL[longer[comb[i]]]) {
					weight++
				}
			}
			if weight > bestWeight {
				bestWeight = weight
				copy(best, comb)
			}
			return
		}
		for i := start; i <= n; i++ {
			comb[k] = i
			combinations(i+1, n, k+1)
		}
	}
	combinations(0, len(longer)-1, 0)
	return best
}

// expandRanges is used for IGNORE_WHITESPACES policy: expand change ranges
// by trimming equal borders.
func expandRanges(lines1, lines2 []*Line, iterable FairDiffIterable) FairDiffIterable {
	var changes []util.Range
	for _, ch := range iterable.Changes() {
		expanded := expandRangeLines(lines1, lines2, ch.Start1, ch.Start2, ch.End1, ch.End2)
		if !expanded.IsEmpty() {
			changes = append(changes, expanded)
		}
	}
	return Fair(CreateFromRanges(changes, len(lines1), len(lines2)))
}

func toLines(text []string, policy ComparisonPolicy) []*Line {
	result := make([]*Line, len(text))
	for i, s := range text {
		result[i] = NewLine(s, policy)
	}
	return result
}

func convertMode(lines []*Line, policy ComparisonPolicy) []*Line {
	result := make([]*Line, len(lines))
	for i, l := range lines {
		if l.Policy != policy {
			result[i] = NewLine(l.Content, policy)
		} else {
			result[i] = l
		}
	}
	return result
}

// MyersMatcher implements LineMatcher using the Myers algorithm with the
// JetBrains two-step comparison approach (smart line comparison + chunk optimization).
type MyersMatcher struct{}

func (m *MyersMatcher) Match(left, right []string, policy ComparisonPolicy) FairDiffIterable {
	return CompareLines(left, right, policy)
}

File: swift/Sources/FileDiff/ByLine.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

/// Compares two sequences of text lines at the line level.
///
/// This is the main entry point of the ByLine comparison stage.
public func compareLines(_ lines1: [String], _ lines2: [String], policy: ComparisonPolicy) -> FairDiffIterable {
    let first = toLines(lines1, policy: policy)
    let second = toLines(lines2, policy: policy)
    return doCompareLines(first, second, policy: policy)
}

/// Line matcher backed by the Myers O(ND) algorithm with JetBrains' two-step refinement.
public struct MyersMatcher: LineMatcher {
    public init() {}

    public func match(_ left: [String], _ right: [String], policy: ComparisonPolicy) -> FairDiffIterable {
        compareLines(left, right, policy: policy)
    }
}

func doCompareLines(_ lines1: [Line], _ lines2: [Line], policy: ComparisonPolicy) -> FairDiffIterable {
    if policy == .ignoreWhitespaces {
        var changes = compareSmart(lines1, lines2)
        changes = optimizeLineChunks(lines1, lines2, changes)
        return expandRanges(lines1, lines2, changes)
    }

    let iwLines1 = convertMode(lines1, policy: .ignoreWhitespaces)
    let iwLines2 = convertMode(lines2, policy: .ignoreWhitespaces)

    var iwChanges = compareSmart(iwLines1, iwLines2)
    iwChanges = optimizeLineChunks(lines1, lines2, iwChanges)
    return correctChangesSecondStep(lines1, lines2, iwChanges)
}

private func compareSmart(_ lines1: [Line], _ lines2: [Line]) -> FairDiffIterable {
    let threshold = unimportantLineCharCount
    if threshold == 0 {
        return diffLineSlice(lines1, lines2)
    }
    let (bigLines1, indexes1) = getBigLines(lines1, threshold: threshold)
    let (bigLines2, indexes2) = getBigLines(lines2, threshold: threshold)
    let changes = diffLineSlice(bigLines1, bigLines2)
    return SmartLineChangeCorrector(
        indexes1: indexes1, indexes2: indexes2,
        lines1: lines1, lines2: lines2,
        changes: changes
    ).build()
}

private func getBigLines(_ lines: [Line], threshold: Int) -> ([Line], [Int]) {
    var big: [Line] = []
    var indexes: [Int] = []
    for (index, line) in lines.enumerated() where line.nonSpaceChars > threshold {
        big.append(line)
        indexes.append(index)
    }
    return (big, indexes)
}

private func expandRanges(_ lines1: [Line], _ lines2: [Line], _ iterable: FairDiffIterable) -> FairDiffIterable {
    var changes: [DiffRange] = []
    for change in iterable.changes {
        let expanded = expandRangeLines(
            lines1, lines2,
            change.left.lowerBound, change.right.lowerBound,
            change.left.upperBound, change.right.upperBound
        )
        if !expanded.isEmpty {
            changes.append(expanded)
        }
    }
    return createFromRanges(changes, length1: lines1.count, length2: lines2.count)
}

// swiftlint:disable function_body_length
private func correctChangesSecondStep(
    _ lines1: [Line],
    _ lines2: [Line],
    _ changes: FairDiffIterable
) -> FairDiffIterable {
    // swiftlint:enable function_body_length
    let eqLines1 = lines1.map { $0 as LineEquatable }
    let eqLines2 = lines2.map { $0 as LineEquatable }
    let builder = ExpandChangeBuilder(objects1: eqLines1, objects2: eqLines2)

    var sample: String?
    var last1 = 0
    var last2 = 0

    for unchanged in changes.unchanged {
        let count = unchanged.left.count
        for i in 0..<count {
            let idx1 = unchanged.left.lowerBound + i
            let idx2 = unchanged.right.lowerBound + i
            let line1 = lines1[idx1]
            let line2 = lines2[idx2]

            if sample == nil || !TextUtils.isEqual(sample ?? "", line1.content, policy: .ignoreWhitespaces) {
                if line1.equals(line2) {
                    flushSecondStep(
                        builder: builder,
                        lines1: lines1,
                        lines2: lines2,
                        sample: sample,
                        last1: last1,
                        last2: last2,
                        line1: idx1,
                        line2: idx2
                    )
                    sample = nil
                    builder.markEqual(idx1, idx2)
                } else {
                    flushSecondStep(
                        builder: builder,
                        lines1: lines1,
                        lines2: lines2,
                        sample: sample,
                        last1: last1,
                        last2: last2,
                        line1: idx1,
                        line2: idx2
                    )
                    sample = line1.content
                }
            }
            last1 = idx1 + 1
            last2 = idx2 + 1
        }
    }
    flushSecondStep(
        builder: builder, lines1: lines1, lines2: lines2,
        sample: sample, last1: last1, last2: last2,
        line1: changes.length1, line2: changes.length2
    )
    return fair(builder.finish())
}

// swiftlint:disable:next function_parameter_count
private func flushSecondStep(
    builder: ExpandChangeBuilder,
    lines1: [Line],
    lines2: [Line],
    sample: String?,
    last1: Int,
    last2: Int,
    line1: Int,
    line2: Int
) {
    guard let sample else {
        return
    }
    let start1 = max(last1, builder.index1)
    let start2 = max(last2, builder.index2)

    var sub1: [Int] = []
    var sub2: [Int] = []
    for i in start1..<line1 where TextUtils.isEqual(sample, lines1[i].content, policy: .ignoreWhitespaces) {
        sub1.append(i)
    }
    for i in start2..<line2 where TextUtils.isEqual(sample, lines2[i].content, policy: .ignoreWhitespaces) {
        sub2.append(i)
    }

    if sub1.isEmpty || sub2.isEmpty {
        return
    }
    alignExactMatching(builder: builder, lines1: lines1, lines2: lines2, sub1: sub1, sub2: sub2)
}

private func alignExactMatching(
    builder: ExpandChangeBuilder,
    lines1: [Line],
    lines2: [Line],
    sub1: [Int],
    sub2: [Int]
) {
    let size = max(sub1.count, sub2.count)
    let skipAligning = size > 10 || sub1.count == sub2.count

    if skipAligning {
        let count = min(sub1.count, sub2.count)
        for i in 0..<count where lines1[sub1[i]].equals(lines2[sub2[i]]) {
            builder.markEqual(sub1[i], sub2[i])
        }
        return
    }

    if sub1.count < sub2.count {
        let matching = getBestMatchingAlignment(shorter: sub1, longer: sub2, linesS: lines1, linesL: lines2)
        for i in 0..<sub1.count where lines1[sub1[i]].equals(lines2[sub2[matching[i]]]) {
            builder.markEqual(sub1[i], sub2[matching[i]])
        }
    } else {
        let matching = getBestMatchingAlignment(shorter: sub2, longer: sub1, linesS: lines2, linesL: lines1)
        for i in 0..<sub2.count where lines1[sub1[matching[i]]].equals(lines2[sub2[i]]) {
            builder.markEqual(sub1[matching[i]], sub2[i])
        }
    }
}

private func getBestMatchingAlignment(shorter: [Int], longer: [Int], linesS: [Line], linesL: [Line]) -> [Int] {
    let size = shorter.count
    var best = Array(0..<size)
    var comb = Array(repeating: 0, count: size)
    var bestWeight = 0

    func combinations(start: Int, end: Int, depth: Int) {
        if depth == size {
            var weight = 0
            for i in 0..<size where linesS[shorter[i]].equals(linesL[longer[comb[i]]]) {
                weight += 1
            }
            if weight > bestWeight {
                bestWeight = weight
                best = comb
            }
            return
        }
        var i = start
        while i <= end {
            comb[depth] = i
            combinations(start: i + 1, end: end, depth: depth + 1)
            i += 1
        }
    }
    combinations(start: 0, end: longer.count - 1, depth: 0)
    return best
}
